### PR TITLE
Forbid incorrectly swapped variables #788

### PR DESCRIPTION
**Pull request by PrVrSs**
# I have made things!

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

### CODE REVIEW

File: wemake_python_styleguide/violations/best_practices.py [hunk 1]
```diff
@@ -1851,3 +1853,37 @@ class ApproximateConstantViolation(ASTViolation):
...
     code = 447
     error_template = 'Found approximate constant: {0}'
+
+
+@final
+class AlmostSwappedViolation(ASTViolation):
+    """
+    Forbids unpythonic swap variables.
+
+    We check for a = b; b = a sequences.
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/best_practices.py]: We use double backquotes to highlight the code. Please, add it to `a = b; b = a`

File: wemake_python_styleguide/visitors/ast/statements.py [hunk 2]
```diff
@@ -149,9 +154,56 @@ def visit_statement_with_body(self, node: StatementWithBody) -> None:
...
         self._check_useless_node(node, node.body)
         self.generic_visit(node)
 
+    def _check_swapped_variables(
+        self,
+        body: Sequence[ast.stmt],
+    ) -> None:
+        for assigns in sequence_of_node((ast.Assign,), body):
+            self._almost_swapped(
+                cast(Sequence[ast.Assign], assigns),
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/statements.py]: Do we need a `cast` here? Why can't we return the correct type from `sequence_of_node`?

File: wemake_python_styleguide/visitors/ast/statements.py [hunk 3]
```diff
@@ -149,9 +154,56 @@ def visit_statement_with_body(self, node: StatementWithBody) -> None:
...
+            )
+
+    def _almost_swapped(self, assigns: Sequence[ast.Assign]) -> None:
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/statements.py]: This method looks very complex, our linter is also marking it like so. Can you please refactor it a bit?

File: wemake_python_styleguide/violations/best_practices.py (hunk 1 above)
Comment (PrVrSs) [reply wemake_python_styleguide/violations/best_practices.py]: done

File: wemake_python_styleguide/visitors/ast/statements.py (hunk 2 above)
Comment (PrVrSs) [reply wemake_python_styleguide/visitors/ast/statements.py]: done

File: wemake_python_styleguide/visitors/ast/statements.py (hunk 3 above)
Comment (PrVrSs) [reply wemake_python_styleguide/visitors/ast/statements.py]: probably done

File: wemake_python_styleguide/logic/collections.py [hunk 4]
```diff
@@ -33,30 +46,33 @@ def normalize_dict_elements(node: ast.Dict) -> Sequence[ast.AST]:
 
 
 def sequence_of_node(
-    node_type: Tuple[Type[ast.stmt]],
+    node_types: Tuple[Type[_NodeType]],
```
Comment (sobolevn) [inline wemake_python_styleguide/logic/collections.py]: It should be `Tuple[Type[_NodeType], ...]`

Comment (PrVrSs) [reply wemake_python_styleguide/logic/collections.py]: done

**Comment by PrVrSs** [comment]
теперь правда, всегда будет по-разному: `a, b = b, a` `|`  `b, a = a, b`. Не могу сказать насколько это критично

**Comment by sobolevn** [comment]
@PrVrSs can we just drop this formatting part? It does not bring much value (because we can use `show-source = true`). Will it make the source code easier? What do you think?

**Comment by sobolevn** [comment]
Thanks a lot for your work. I really appreciate that. 👍

**Pull request merged**